backend/fbdev.c

```diff
@@ -0,0 +1,315 @@
+
+    if (twin_screen_damaged(screen))
+        twin_screen_update(screen);
+    return true;
+}
+
+static bool twin_fbdev_apply_config(twin_fbdev_t *tx)
+{
+    /* Read changable information of the framebuffer */
+    if (ioctl(tx->fb_fd, FBIOGET_VSCREENINFO, &tx->fb_var) == -1) {
+        printf("error : failed getting framebuffer information\n");
+        return false;
+    }
+
+    /* Set the virtual screen size to be the same as the physical screen */
+    tx->fb_var.xres_virtual = tx->fb_var.xres;
+    tx->fb_var.yres_virtual = tx->fb_var.yres;
+    tx->fb_var.bits_per_pixel = 32;
+    if (ioctl(tx->fb_fd, FBIOPUT_VSCREENINFO, &tx->fb_var) < 0) {
+        printf("Failed to set framebuffer mode\n");
+        return false;
+    }
+
+    /* Read changable information of the framebuffer again */
+    if (ioctl(tx->fb_fd, FBIOGET_VSCREENINFO, &tx->fb_var) < 0) {
+        printf("Failed to get framebuffer information\n");
+        return false;
+    }
+
+    /* Check bits per pixel */
+    if (tx->fb_var.bits_per_pixel != 32) {
+        printf("Failed to set framebuffer bpp to 32\n");
+        return false;
+    }
+
+    /* Read unchangable information of the framebuffer */
+    ioctl(tx->fb_fd, FBIOGET_FSCREENINFO, &tx->fb_fix);
+
+    /* Align the framebuffer memory address with the page size */
+    off_t pgsize = getpagesize();
+    off_t start = (off_t) tx->fb_fix.smem_start & (pgsize - 1);
+
+    /* Round up the framebuffer memory size to match the page size */
+    tx->fb_len = start + (size_t) tx->fb_fix.smem_len + (pgsize - 1);
+    tx->fb_len &= ~(pgsize - 1);
+
+    /* Map framebuffer device to the virtual memory */
+    tx->fb_base = mmap(NULL, tx->fb_len, PROT_READ | PROT_WRITE, MAP_SHARED,
+                       tx->fb_fd, 0);
+    if (tx->fb_base == MAP_FAILED) {
+        printf("Failed to mmap framebuffer.\n");
+        return false;
+    }
+
+    return true;
+}
+
+static bool twin_find_free_vt(int *vt_num)
+{
+    char vt_path[30];
+    int fd;
+
+    /* Find a usable virtual terminal. Since VT1 is occupied by the GNU/Linux
+     * desktop environment and VT2 is unusable, the search begins from VT3.
+     */
+    for (int i = 3; i <= VT_MAX; i++) {
+        snprintf(vt_path, sizeof(vt_path), "/dev/tty%d", i);
+        fd = open(vt_path, O_RDWR);
+        if (fd > 0) {
+            close(fd);
+            *vt_num = i;
+            return true;
+        }
+    }
+
+    /* No free VT found */
+    return false;
+}
+
+static bool twin_vt_setup(twin_fbdev_t *tx)
+{
+    /* Open VT0 to inquire information */
+    tx->vt_fd = open("/dev/tty0", O_RDWR);
+    if (tx->vt_fd == -1) {
+        printf("Failed to open /dev/tty0.\n");
+        return false;
+    }
+
+    /* Find next free virtual terminal */
+    if (!twin_find_free_vt(&tx->vt_num)) {
+        printf("Failed to find a free virtual terminal.\n");
+        return false;
+    }
+    close(tx->vt_fd);
+
+    /* Open the found free virtual terminal */
```

Comment:
Can you avoid duplicating code for accessing tty devices?

Reply:
This is intended for the following manner:
https://github.com/g0hl1n/psplash/blob/master/psplash-console.c#L119

The reason I didn't follow the code in the first place is that the `ioctl()` returns 1 (i.e., VT1) to me, and I thought VT1 is occupied by the desktop graphics environment.